vtrepairtable.go

```diff
@@ -0,0 +1,308 @@
+		log.Fatal(err)
+	}
+
+	if sourceTarget.Keyspace != destTarget.Keyspace {
+		log.Fatal("can only move between shards in the same keyspace")
+	}
+
+	tableName := args[2]
+
+	ts := topo.Open()
+
+	vschema, err := loadVSchema(ctx, ts, sourceTarget.Keyspace)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	sourceQs, _, err := Connect(ctx, ts, sourceTarget)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	destShard, err := ts.GetShard(ctx, destTarget.Keyspace, destTarget.Shard)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	destQs, _, err := Connect(ctx, ts, destTarget)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	sourceTx, err := sourceQs.Begin(ctx, sourceTarget, nil)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer maybeRollback(sourceQs, sourceTarget, sourceTx)
+
+	destTx, err := destQs.Begin(ctx, destTarget, nil)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer maybeRollback(destQs, destTarget, destTx)
+
+	selectSql := fmt.Sprintf("SELECT * FROM %v", tableName)
+	log.Infoln(selectSql)
+
+	response, err := sourceQs.Execute(ctx, sourceTarget, selectSql, nil, sourceTx, nil)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var fieldVars = make([]string, len(response.Fields))
+	for i, field := range response.Fields {
+		fieldVars[i] = ":" + field.Name
+	}
+	insertSql := "INSERT INTO " + sqlescape.EscapeID(tableName) +
+		" (" + strings.Join(escapeAll(response.Fields), ", ") + ") VALUES (" +
+		strings.Join(fieldVars, ",") + ")"
+	// TODO handle primary keys properly
+	deleteSql := "DELETE FROM " + sqlescape.EscapeID(tableName) + " WHERE id = ?"
+
+	table := vschema.Tables[tableName]
+	primaryVindex := table.ColumnVindexes[0]
+	if len(primaryVindex.Columns) > 1 {
+		log.Fatal("can only handle vindex on single column")
+	}
+
+	uniqueVindex, ok := primaryVindex.Vindex.(vindexes.Unique)
+	if !ok {
+		log.Fatal("primary vindex have to be unqiue")
+	}
+
+	// TODO handle primary keys properly
+	idColumnIndex := -1
+	primaryVindexedColumn := primaryVindex.Columns[0].Lowered()
+	primaryVindexColumIndex := -1
+	for i, field := range response.Fields {
+		if field.Name == "id" {
+			idColumnIndex = i
+		}
+		if field.Name == primaryVindexedColumn {
+			primaryVindexColumIndex = i
+		}
+	}
+	if idColumnIndex == -1 {
+		log.Fatalf("no column named 'id' in: %#v", response.Fields)
+	}
+	if primaryVindexColumIndex == -1 {
+		log.Fatalf("could not find primary vindexed column '%v' in: %#v", primaryVindexedColumn, response.Fields)
+	}
+
+	log.Infof("found %v rows", len(response.Rows))
+	for _, row := range response.Rows {
+		// TODO TODO TODO TODO TODO TODO TODO
+		// Check that the row actually belongs to the destination shard
+		primaryVindexVal := row[primaryVindexColumIndex]
+		ksids, err := uniqueVindex.Map(nil, []sqltypes.Value{primaryVindexVal})
+		if err != nil {
+			log.Fatal(err)
+		}
+		if ksids[0].Range != nil || ksids[0].ID == nil {
+			log.Fatalf("could not map %v to a keyspace id", primaryVindexVal)
+		}
+		if !key.KeyRangeContains(destShard.KeyRange, ksids[0].ID) {
```

Comment:
Why not go through and come up with a mapping from id on the wrong shard to the correct shard.  Then divide the ids to move by shard and then open a transaction on each shard at that time?